feat: add functionality to support custom tokens

mikasackermn opened the pull request:
# Summary

Users can add tokens that aren't automatically included in our default token list by using the custom token functionality.

- When a custom token is added, it should be displayed on all related pages, including the token selector.
- The custom token should be treated like any other token from the meta-response token list.
- Ensure the balance of the custom token is displayed correctly.
- Implement for EVM and Solana blockchains.(The blockchain selector should be limited to these blockchains.)
- Custom tokens should be stored in Zustand (local storage).
- Tokens should be sorted based on the user’s insertion time, with recently added tokens shown on top.
- When the API malfunctions or the token data is inaccurate, present a mo- When the API malfunctions or the token data is inaccurate, present a modal in line with the design.
- Show an error message if a user tries to import a token that has already been added or exists in the meta.dal in line with the design.
- Show an error message if a user tries to import a token that has already been added or exists in the meta.
- Add a boolean to the widget configuration to allow hiding custom tokens from the settings.
Fixes # (issue)


# How did you test this change?

By choosing custom token on the settings page, you may test it.

Sample test case:
- 3yoMkf3X6bDxjks6YaWwNk4SAbuaysLg1a4BjQKToQAA


# Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Implemented a user interface (UI) change, referencing our Figma design to ensure pixel-perfect precision.

RyukTheCoder commented on widget/ui/src/components/NotFound/NotFound.tsx:
```diff
@@ -9,9 +9,10 @@ import { Typography } from '../Typography';
   return (
     <Container>
-      <SearchIcon color="secondary" size={26} />
+      {hasIcon && <SearchIcon color="secondary" size={26} />}
```
Maybe it is better to receive the icon as prop and show the default icon if it is not available.

mikasackermn replied:
You're right, fixed

RyukTheCoder commented on widget/embedded/src/store/slices/settings.ts:
```diff
@@ -131,6 +139,16 @@ export const createSettingsSlice: StateCreator<
     set((state) => ({
       infiniteApprove: !state.infiniteApprove,
     })),
+  setCustomTokens: (token) =>
```
A token is being added to the list of custom tokens so maybe `setCustomTokens` name is not suitable for this.

mikasackermn replied:
I change to setCustomToken

RyukTheCoder replied:
Maybe 'add' or 'insert' is better

mikasackermn replied:
Thank you , fixed

RyukTheCoder replied:
I think it is still 'setCustomToken'

RyukTheCoder commented on widget/embedded/src/store/slices/data.ts:
```diff
@@ -117,6 +139,9 @@ export const createDataSlice: StateCreator<
       cacheService.set(cacheKey, supportedTokens);
     }
 
+    supportedTokens = supportedTokens.concat(customTokens);
```
Shouldn't we check for customToken in config to see if we should concat customTokens?
Also please remove the console log.

mikasackermn replied:
fixed

RyukTheCoder reviewed (changes requested):
Tnx, I just added some comments.

mikasackermn commented on widget/embedded/src/pages/AddCustomTokenPage.tsx:
```diff
@@ -0,0 +1,189 @@
+              {i18n.t('Enter Address')}
+            </Typography>
+            <Divider size={10} />
+            <TextField
```
fixed

mikasackermn commented on widget/embedded/src/components/CustomTokenModal/CustomTokenModal.tsx:
```diff
@@ -0,0 +1,86 @@
+        </Typography>
+        <Divider size={4} />
+        <Typography variant="body" size="medium">
+          <StyledLink
```
fixed

mikasackermn commented on widget/embedded/src/pages/CustomTokensPage.tsx:
```diff
@@ -0,0 +1,154 @@
+import { useAppStore } from '../store/AppStore';
+import { getContainer } from '../utils/common';
+
+const Content = styled('div', {
```
fixed

RyukTheCoder replied:
It seems the gap did not applied

![image](https://github.com/user-attachments/assets/5cb8c531-854c-45f8-b661-c38b9a1e6c4f)

mikasackermn replied:
fixed

mikasackermn commented on widget/embedded/src/pages/CustomTokensPage.tsx:
```diff
@@ -0,0 +1,154 @@
+              />
+              <Divider size={16} />
+              <TokenList
+                list={customTokens}
```
fixed

mikasackermn commented on widget/embedded/src/pages/CustomTokensPage.tsx:
```diff
@@ -0,0 +1,154 @@
+  const { customTokens, deleteCustomToken, theme } = useAppStore();
+  const isDarkTheme = theme === 'dark';
+  const navigate = useNavigate();
+  const [open, setOpen] = useState<boolean>(false);
```
I changed to isOpenDeleteModal because I think it makes more sense.

mikasackermn commented on widget/embedded/src/pages/AddCustomTokenPage.tsx:
```diff
@@ -0,0 +1,189 @@
+    setCustomTokens,
+  } = useAppStore();
+  const [address, setAddress] = useState('');
+  const [openErrorModal, setOpenErrorModal] = useState<boolean>(false);
```
I changed to isOpenErrorModal

mikasackermn commented on widget/embedded/src/pages/AddCustomTokenPage.tsx:
```diff
@@ -0,0 +1,189 @@
+
+  const isValidAddress =
+    !!selectedBlockchainForCustomToken &&
+    ((selectedBlockchainForCustomToken.type === TransactionType.SOLANA &&
```
fixed

mikasackermn commented on widget/embedded/src/hooks/useFetchCustomToken.ts:
```diff
@@ -0,0 +1,107 @@
+    setError({
+      title: i18n.t('Token Not Found'),
+      message: i18n.t(
+        'Sorry, no token was found on Binance blockchain with the provided address. please make sure you have entered the right token address.'
```
fixed

mikasackermn commented on widget/embedded/src/components/TokenList/TokenListContent.tsx:
```diff
@@ -0,0 +1,21 @@
+/* eslint-disable @typescript-eslint/no-magic-numbers */
```
I removed this file, please check

RyukTheCoder commented on widget/embedded/src/pages/SelectSwapItemsPage.tsx:
```diff
@@ -19,6 +19,13 @@ interface PropTypes {
   type: 'source' | 'destination';
 }
 
+export const TokenContainer = styled('div', {
```
```suggestion
export const TokenListContainer = styled('div', {
```

mikasackermn replied:
fixed

RyukTheCoder commented on widget/embedded/src/components/TokenList/TokenList.tsx:
```diff
@@ -112,151 +117,146 @@ export function TokenList(props: PropTypes) {
-          const customCssForTagTitle = {
-            $$color: colors.main750,
-            [`.${darkTheme} &`]: {
+      <List as="ul">
```
'ul' element should only contain 'li' elements.

mikasackermn replied:
fixed

RyukTheCoder commented on widget/embedded/src/components/TokenList/TokenList.tsx:
```diff
@@ -112,151 +117,146 @@ export function TokenList(props: PropTypes) {
+    <LoadingTokenList size={PAGE_SIZE} />
+  ) : (
+    fetchStatus === 'success' && (
+      <List as="ul">
```
'ul' element should only contain 'li' elements.

mikasackermn replied:
This change has already been made and isn't related to this PR. Additionally, it's possible that modifying this part of VirtualizedList may not work correctly.

RyukTheCoder reviewed (changes requested):
Tnx, I just added some comments.

mikasackermn commented on widget/embedded/src/components/CustomTokenModal/CustomTokenModal.tsx:
```diff
@@ -33,32 +33,39 @@ export function CustomTokenModal(props: PropTypes) {
-            <ExternalLinkIcon size={12} color="gray" />
-          </StyledLink>
+          {token?.address &&
+          blockchain?.type !== 'COSMOS' &&
```
fixed, I added a comment

mikasackermn commented on widget/embedded/src/components/CustomTokenModal/CustomTokenModal.tsx:
```diff
@@ -33,32 +33,39 @@ export function CustomTokenModal(props: PropTypes) {
+          !!blockchain?.info?.addressUrl ? (
+            <StyledLink
+              hasHover
+              href={blockchain.info.addressUrl
```
fixed

RyukTheCoder reviewed (approved):
LGTM

yeager-eren commented on widget/embedded/src/store/slices/data.ts:
```diff
@@ -68,6 +79,14 @@ export const createDataSlice: StateCreator<
+    if (options.type === 'custom') {
+      return blockchainsFromState.filter(
+        (blockchain) =>
+          blockchain.type === TransactionType.EVM ||
```
Please create a const called `ACTIVE_BLOCKCHAINS_FOR_CUSTOM_TOKENS` and put it outside of this method to clearly communicate the logic and also simple add more blockchains in future.

yeager-eren commented on widget/embedded/src/store/slices/data.ts:
```diff
@@ -1,21 +1,32 @@
 
 type BlockchainOptions = {
-  type?: 'source' | 'destination';
+  type?: 'source' | 'destination' | 'custom';
```
```suggestion
  type?: 'source' | 'destination' | 'custom-tokens';
```

yeager-eren commented on widget/embedded/src/pages/SelectSwapItemsPage.tsx:
```diff
@@ -95,24 +102,32 @@ export function SelectSwapItemsPage(props: PropTypes) {
+          <Typography variant="label" size="large">
+            {i18n.t('Select Token')}
+          </Typography>
+          <Divider size={4} />
```
```suggestion
          <Divider size={!tokens.length && !!searchedFor ? 36 : 4} />
```

yeager-eren commented on widget/embedded/src/pages/SelectBlockchainPage.tsx:
```diff
@@ -24,15 +25,17 @@ export function SelectBlockchainPage(props: PropTypes) {
 
-  const showCategory = activeCategoriesCount !== 1;
+  const showCategory =
+    props.showCategory === false
```
You can change `showCategory` to `hideCategory` in your props. and update this code to something like this:
```
const showCategory = props.hideCategory : false : activeCategoriesCount !== 1;
```

yeager-eren commented on widget/embedded/src/pages/HistoryPage.tsx:
```diff
@@ -186,7 +186,7 @@ export function HistoryPage() {
                 }
                 titleColor={!searchedFor ? '$info' : undefined}
-                hasIcon={!!searchedFor}
+                icon={!searchedFor && <></>}
```
What is this for?

yeager-eren commented on widget/embedded/src/pages/CustomTokensPage.tsx:
```diff
@@ -0,0 +1,165 @@
+    setSearchedFor(value);
+  };
+
+  const filterCustomTokens = () =>
```
There is no point to make this lazy, because you are calling it in your render always.

yeager-eren replied:
I think it's not needed to use `useMemo` here since `searchedFor` and `customTokens` are the main values of this page which means they will cause rerender in most cases. And also custom tokens wouldn't be a long list I guess, so memoizing the computation wouldn't help us to much. I prefer to keep the code simple and use this types of optimization only if needed.

yeager-eren replied:
`filter` also is a verb which can be used for actions/functions. here you are keeping result of your filter. I suggest `filteredCustomTokens` or a more contextual name to show search functionality like `customTokensResults`. I prefer the latter one.

yeager-eren replied:
There is also a `console` at line 69. please remove.

mikasackermn replied:
Okay, I changed the variable name to ### customTokensResults and returned the function to simple code. I also removed the console.

yeager-eren commented on widget/embedded/src/components/BlockchainList/BlockchainList.tsx:
```diff
@@ -18,7 +18,13 @@ import { Container, List } from './BlockchainList.styles';
+    searchedFor,
+    onChange,
+    blockchainCategory,
+    showLabel = true,
```
it's not `label` to me. `title` is more suitable I guess.

yeager-eren reviewed (changes requested):
I reviewed some of your changes. will continue my review soon. You can check my comments in between.

mikasackermn commented on widget/embedded/src/pages/SelectBlockchainPage.tsx:
```diff
@@ -14,7 +14,8 @@ import { useAppStore } from '../store/AppStore';
 
 interface PropTypes {
-  type: 'source' | 'destination';
+  type: 'source' | 'destination' | 'custom';
```
fixed

mikasackermn commented on widget/ui/src/icons/index.ts:
```diff
@@ -12,7 +12,10 @@ export { default as CompleteIcon } from './Complete';
 export { default as CopyIcon } from './Copy';
 export { default as CosmosCategoryIcon } from './CosmosCategory';
 export { default as CreditCardIcon } from './CreditCard';
+export { default as CustomTokensIcon } from './CustomTokens';
```
fixed

mikasackermn commented on widget/ui/svgs/resources/fill/Custom Tokens.svg:
fixed

mikasackermn commented on widget/ui/src/icons/CustomTokens.tsx:
```diff
@@ -0,0 +1,376 @@
+      viewBox="0 0 200 173"
+      fill="none"
+      xmlns="http://www.w3.org/2000/svg">
+      <g id="Custom Tokens">
```
fixed

yeager-eren commented on widget/embedded/src/components/ItemPicker/ItemPicker.tsx:
```diff
@@ -0,0 +1,57 @@
+  );
+}
+
+export default React.memo(ItemPicker);
```
We don't use `default` exports.

mikasackermn replied:
Fixed it

yeager-eren commented on widget/embedded/src/components/ItemPicker/ItemPicker.tsx:
```diff
@@ -0,0 +1,57 @@
+    disabled,
+  } = props;
+
+  const LogoComponent = logo;
```
Seems redundant.

mikasackermn replied:
Fixed it

yeager-eren commented on widget/embedded/src/components/ItemPicker/ItemPicker.tsx:
```diff
@@ -0,0 +1,57 @@
+
+import { Container, FlexContainer, InputContainer } from './ItemPicker.styles';
+
+function ItemPicker(props: PropTypes) {
```
`Item` is too general for this component, and also `picker` is suitable for cases you are picking a value from a list that passed to component.

yeager-eren replied:
I suggest `BlockchainSelectorButton` or something like this.

mikasackermn replied:
fixed

yeager-eren commented on widget/embedded/src/components/ItemPicker/ItemPicker.types.ts:
```diff
@@ -0,0 +1,11 @@
+  onClick: () => void;
+  value: {
+    label?: string;
+    logo?: string | React.ComponentType<{ size?: number }>;
```
You never used `logo` with a react component. I don't know why you've added to the interface.

mikasackermn replied:
fixed

yeager-eren commented on widget/embedded/src/components/ItemPicker/ItemPicker.types.ts:
```diff
@@ -0,0 +1,11 @@
+export interface PropTypes {
+  onClick: () => void;
+  value: {
```
Type can be improve by changing it something like this:

```js
  value?: {
    name: string, // this filed can be `displayName` or `title` as well.
    logo: string,
  }
```

mikasackermn replied:
fixed

yeager-eren replied:
Remove my comment please. I wrote the comment to you.

mikasackermn replied:
I Removed it

yeager-eren commented on widget/embedded/src/hooks/useFetchCustomToken.ts:
```diff
@@ -0,0 +1,107 @@
+  const [error, setError] = useState<undefined | ErrorType>(undefined);
+  const { findToken, customTokens } = useAppStore();
+
+  const handleTokenIsDuplicateError = () => {
```
Instead of this approach you could do this:

```
function produceErrorMessage(type: "duplicated" | "not-found" | ...){
switch(type){
case "...":
 return {
   title: ...,
   message: ..,
   type: ...
 }
break;
}
}
```

and usage will be like: 

```
      if (isDuplicate) {
        return setError(produceErrorMessage("duplicated"));
      }
```

mikasackermn replied:
fixed

yeager-eren commented on widget/embedded/src/hooks/useFetchCustomToken.ts:
```diff
@@ -0,0 +1,107 @@
+
+      return response;
+    } catch (error: any) {
+      if (error.message === 'Failed to fetch') {
```
You don't need to detect network like this. I suggest to completely remove the need of `type` which can `NETWORK_ERROR` | `TOKEN_ERROR`. Let this method `fetchCustomToken` throw an error (don't catch it), so you will handle network error on where this method is going to use. e.g. `AddCustomTokenPage`.

It will also simplify your code here.

mikasackermn replied:
fixed

yeager-eren commented on widget/embedded/src/components/CustomTokenModal/CustomTokenModal.tsx:
```diff
@@ -0,0 +1,100 @@
+          variant="body"
+          className="_custom-token-description">
+          {i18n.t(
+            'This token doesn’t appear on the active token list(s). Make sure this is the token that you want to trade.'
```
invalid character
```suggestion
            'This token doesn't appear on the active token list(s). Make sure this is the token that you want to trade.'
```

mikasackermn replied:
fixed

yeager-eren commented on widget/embedded/src/components/CustomTokenModal/CustomTokenModal.tsx:
```diff
@@ -0,0 +1,100 @@
+              className="_coin-source-name"
+              variant="body"
+              size="xsmall">
+              {token?.coinSource}
```
```suggestion
              {token.coinSource}
```

yeager-eren commented on widget/embedded/src/components/CustomTokenModal/CustomTokenModal.types.ts:
```diff
@@ -0,0 +1,9 @@
+  open: boolean;
+  onClose: () => void;
+  handleSubmitClick: () => void;
+  token?: Token;
```
Why this is optional?

mikasackermn replied:
I fixed it.
I use this condition `selectedBlockchainForCustomToken && token` to open modals.

yeager-eren commented on widget/embedded/src/components/CustomTokenModal/CustomTokenModal.tsx:
```diff
@@ -0,0 +1,100 @@
+export function CustomTokenModal(props: PropTypes) {
+  const { open, onClose, token, handleSubmitClick, blockchain } = props;
+
+  const shouldShowAddressLink =
```
Please replace `shouldShow...` and `getFormatted...` with following changes.

Create a `CustomTokenModal.helpers.ts` then:
```

export function generateExplorerLink(
  address: string,
  blockchain: BlockchainMeta
): string {
  // We don't have Cosmos explorer url in API.
  if (blockchain.type === 'COSMOS') {
    return '';
  }

  const explorerLinkPattern = blockchain.info?.addressUrl;

  if (!explorerLinkPattern) {
    return '';
  }

  return explorerLinkPattern.replace('{wallet}', address);
};
```

Then, in this component:


```
export function CustomTokenModal(props: PropTypes) {
  // ...
   const explorerLink = generateExplorerLink(token.address);
   // ..
}
```

mikasackermn replied:
fixed

yeager-eren commented on widget/embedded/src/pages/AddCustomTokenPage.tsx:
```diff
@@ -0,0 +1,192 @@
+            </Button>
+          </MessageBox>
+        </WatermarkedModal>
+        {!!selectedBlockchainForCustomToken && (
```
I have two question here:
1. Why your modal is condional?
2. If it should be, why you didn't use `isOpenImportModa` instead of `selectedBlockchainForCustomToken` for your condition?

mikasackermn replied:
There is no need to have a condition, I fixed it

mikasackermn replied:
Since I wanted the `Blockchain` to be mandatory in the modal and the first `selectedBlockchainForCustomToken` is empty

mikasackermn replied:
I deleted `isOpenImportModal`

yeager-eren reviewed (changes requested):
I've put some comments on your PR. I think after resolving the comment, it will need another quick review. I also didn't review your apporach on`widget/ui/src/icons/CustomTokensZeroState.tsx` yet. will do soon.

mikasackermn commented on widget/embedded/src/components/ItemPicker/ItemPicker.types.ts:
```diff
@@ -0,0 +1,11 @@
+    logo?: string | React.ComponentType<{ size?: number }>;
+  };
+  title: string;
+  placeholder?: string;
```
fixed

mikasackermn commented on widget/embedded/src/components/TokenList/TokenList.tsx:
```diff
@@ -112,151 +117,145 @@ export function TokenList(props: PropTypes) {
-        </>
-      );
-    }
+  if (!tokens.length && !!searchedFor) {
```
I understand, I revert back the previous code and add our needed change to the code

mikasackermn commented on widget/embedded/src/components/TokenList/TokenList.tsx:
```diff
@@ -91,14 +90,20 @@ const renderDesc = (props: RenderDescProps) => {
+    searchedFor = '',
+    onChange,
+    selectedBlockchain,
+    endTokensItem,
```
I changed name to action

mikasackermn commented on widget/embedded/src/hooks/useTheme.ts:
```diff
@@ -93,8 +93,14 @@ export function useTheme(props: WidgetTheme) {
+    return effectiveTheme === 'dark';
+  };
+
+  return { activeTheme: getActiveTheme, isDarkTheme: isDarkTheme() };
```
fixed

mikasackermn commented on widget/embedded/src/pages/CustomTokensPage.tsx:
```diff
@@ -0,0 +1,169 @@
+  const navigate = useNavigate();
+  const [isOpenDeleteModal, setIsOpenDeleteModal] = useState<boolean>(false);
+  const [selectedToken, setSelectedToken] = useState<Token>();
+  const searchHandler = (event: React.ChangeEvent<HTMLInputElement>) => {
```
fixed

mikasackermn commented on widget/embedded/src/pages/CustomTokensPage.tsx:
```diff
@@ -0,0 +1,169 @@
+  const { customTokens, deleteCustomToken } = useAppStore();
+  const { isDarkTheme } = useTheme({});
+  const navigate = useNavigate();
+  const [isOpenDeleteModal, setIsOpenDeleteModal] = useState<boolean>(false);
```
fixed

mikasackermn commented on widget/embedded/src/store/slices/settings.ts:
```diff
@@ -25,7 +25,10 @@ export interface SettingsSlice {
+  customTokens: Token[];
+  selectedBlockchainForCustomToken: BlockchainMeta | null;
 
+  setSelectedBlockchainForCustomToken: (chian: BlockchainMeta | null) => void;
```
fixed

mikasackermn commented on widget/embedded/src/components/TokenList/TokenList.styles.ts:
```diff
@@ -48,13 +48,6 @@ export const tokenAddressStyles = css({
 export const tokenWithoutNameStyles = css({});
 export const usdValueStyles = css();
 
-export const Container = styled('div', {
```
fixed.
I reverted this in TokensList

mikasackermn replied:
I revert back `Container` . Yes, I just added `justifyContent`, because I wanted to display the not found message in the middle of the screen when the token was not found

yeager-eren replied:
You only renamed the component. please also move it on top of `export const Title = styled('div', {` as well.

You can see your change on Github to see where it was and where you moved it.

mikasackermn replied:
fixed

mikasackermn commented on widget/embedded/src/hooks/useFetchCustomToken.ts:
```diff
@@ -0,0 +1,107 @@
+        symbol: response.symbol,
+      });
+      if (isTokenFound) {
+        handleTokenIsDuplicateError();
```
fixed

mikasackermn commented on widget/embedded/src/utils/meta.ts:
```diff
@@ -59,3 +67,20 @@ export function isTokenNative(
+  return type in TOKEN_ADDRESS_PATTERNS;
+}
+
+export function isValidTokenAddress(
```
fixed

yeager-eren replied:
Not sure what was your assumption that you only checking the first pattern when it's actually a list of patterns.
Besides that, we already have something in `widget/embedded/src/components/ConfirmWalletsModal/ConfirmWallets.helpers.ts`. reuse the code by moving that function in utils and update your code and usage of the old code.

Ikari-Shinji-re commented on widget/embedded/src/pages/AddCustomTokenPage.tsx:
```diff
@@ -0,0 +1,208 @@
+  const navigate = useNavigate();
+  const [searchParams] = useSearchParams();
+  const navigateBack = useNavigateBack();
+  console.log('state: ', searchParams.get(SearchParams.BLOCKCHAIN));
```
Please remove logs

mikasackermn replied:
fixed

mikasackermn commented on widget/embedded/src/store/slices/settings.ts:
```diff
@@ -177,4 +195,9 @@ export const createSettingsSlice: StateCreator<
       }),
     });
   },
+  setSelectedBlockchainForCustomToken: (chain) => {
```
fixed

mikasackermn commented:
- Added generateExplorerLink function to get link token
- Added produceErrorMessage function to get custom tokens error
- Handled network error on add custom token page
- Added customTokens to settings store to get available customTokens according to supported blockchains
- Added getSupportedBlockchainsFromConfig function to get similar blockchains between 'from' and 'to' config

yeager-eren commented on widget/embedded/src/constants/searchParams.ts:
```diff
@@ -5,6 +5,7 @@ export enum SearchParams {
   TO_TOKEN = 'toToken',
   FROM_AMOUNT = 'fromAmount',
   AUTO_CONNECT = 'autoConnect',
+  BLOCKCHAIN = 'blockchain',
```
Add a comment that this is for custom tokens.

mikasackermn replied:
I added a comment

yeager-eren commented on widget/embedded/src/store/utils/data.ts:
```diff
@@ -115,3 +119,16 @@ export function matchTokensFromConfigWithMeta(params: {
+  const configFromBlockchains = config.from?.blockchains || [];
+  const configToBlockchains = config.to?.blockchains || [];
+
+  return configFromBlockchains.filter((type) =>
```
You can use `Set` to have unique values.

mikasackermn replied:
I used set for this

yeager-eren replied:
Regarding you change, I think this is redundant and can be removed.
```
  if (!configFromBlockchains.length || !configToBlockchains.length) {
    return [];
  }
```

yeager-eren commented on widget/embedded/src/store/slices/data.ts:
```diff
@@ -67,8 +81,31 @@ export const createDataSlice: StateCreator<
+      });
+
+      // Filter blockchains from state based on common types and active custom tokens
+      return blockchainsFromState.filter((blockchain) => {
```
This filter seems confusing to me. I thought it should be something like:
```javascript
let supportedBlockchainsForCustomTokens: string[] =
        ACTIVE_BLOCKCHAINS_FOR_CUSTOM_TOKENS;

      /*
       * Supported blockchains can be configured and be limited.
       * In this case, we only keep those active blockchains which exist in config.
       */
      if (supportedBlockchainsFromConfig.length > 0) {
        supportedBlockchainsForCustomTokens =
          supportedBlockchainsFromConfig.filter((blockchain) =>
            ACTIVE_BLOCKCHAINS_FOR_CUSTOM_TOKENS.includes(blockchain)
          );
      }

      return blockchainsFromState.filter((blockchain) =>
        supportedBlockchainsForCustomTokens.includes(blockchain.type)
      );
```

to make this work you need to update `widget/embedded/src/constants/customTokens.ts`:
```
export const ACTIVE_BLOCKCHAINS_FOR_CUSTOM_TOKENS: string[] = ...
```

mikasackermn replied:
I changed my code to your suggested code
But for this, we have to store the names of the blockchains in ACTIVE_BLOCKCHAINS_FOR_CUSTOM_TOKES instead of typing them.
Did I understand correctly?

yeager-eren replied:
No. I didn't mean to change `blockchain.type` to `blockchain.name`.  By using my code you will go through a typescript error regarding `ACTIVE_BLOCKCHAINS_FOR_CUSTOM_TOKENS` which should be `TransactionType`. I said to fix this issue you can use: 
```
export const ACTIVE_BLOCKCHAINS_FOR_CUSTOM_TOKENS: string[] = [
TransactionType.EVM,
  TransactionType.SOLANA,
]
```

yeager-eren replied:
Please revert back the code and only add `: string[]` to the const.

mikasackermn replied:
I understood. So, I corrected a minor issue in your suggested code to get `ACTIVE_BLOCKCHAINS_FOR_CUSTOM_TOKENS` back to how it was.

yeager-eren commented on widget/embedded/src/store/slices/settings.ts:
```diff
@@ -177,4 +185,26 @@ export const createSettingsSlice: StateCreator<
+  customTokens: () => {
+    const customTokens = get()._customTokens;
+    const config = get().config;
+    const supportedBlockchainsFromConfig = getSupportedBlockchainsFromConfig({
```
Seems you are repeating `data.ts`'s `blockchains` method here. Couldn't you use that method here? `.blockchains({type: "custom-tokens"})`

mikasackermn replied:
Yes, I used that method here.

mikasackermn commented on widget/embedded/src/components/CustomTokenModal/CustomTokenModal.types.ts:
```diff
@@ -0,0 +1,9 @@
+export type PropTypes = {
+  open: boolean;
+  onClose: () => void;
+  handleSubmitClick: () => void;
```
I changed it to `onSubmitClick`

mikasackermn commented on widget/embedded/src/components/TokenList/TokenList.styles.ts:
```diff
@@ -6,6 +6,14 @@ import {
   Typography,
 } from '@rango-dev/ui';
 
+export const TokenListContainer = styled('div', {
```
I revert back Container . Yes, I just added justifyContent, because I wanted to display the not found message in the middle of the screen when the token was not found

mikasackermn commented on widget/embedded/src/components/TokenList/TokenList.tsx:
```diff
@@ -121,148 +128,157 @@ export function TokenList(props: PropTypes) {
-          const customCssForTagTitle = {
-            $$color: colors.main750,
-            [`.${darkTheme} &`]: {
+      <List as="ul">
```
This was purposefully moved here because the message `not found ` shouldn't be in this tag so that it can be shown in the center of the screen.

yeager-eren replied:
You can change your code to:

```
const hasResult = tokens.length > 0 && !!searchedFor;

// ... 

{fetchStatus === 'success' && hasResult ? <List> renderList() </List>) : <NotFound />}
```

mikasackermn commented on widget/embedded/src/constants/customTokens.ts:
```diff
@@ -0,0 +1,7 @@
+import { TransactionType } from 'rango-types';
+
+// Define the active blockchains for custom tokens
```
I removed this comment

yeager-eren commented on widget/embedded/src/components/TokenList/TokenList.types.ts:
```diff
@@ -1,11 +1,14 @@
+  onChange?: (token: Token) => void;
   selectedBlockchain?: string;
-  type: 'source' | 'destination';
+  type?: 'source' | 'destination';
```
You broke the pinned token functionality. in your previous commit you had a check: `props.type && isTokenPinned(token, props.type) &&`
Now you've changed it to `isTokenPinned(token, props.type) &&` which means `custom-token` type will be passed to this as well. which is not intended.

mikasackermn replied:
I fixed it by  this condition `props.type !== 'custom-token' &&`

yeager-eren commented on widget/embedded/src/components/TokenList/TokenList.tsx:
```diff
@@ -220,7 +217,9 @@ export function TokenList(props: PropTypes) {
                 }
                 end={
-                  loadingWallet ? (
+                  action ? (
```
I'm against of nested ternaries. You can do something like:

```

        const endRenderer = () => {
          if (action) {
            return action(token);
          }
          if (loadingWallet) {
            return (
              <End>
                <Skeleton variant="text" size="large" width={70} />
                <Divider size={4} />
                <Skeleton variant="text" size="medium" width={50} />
              </End>
            );
          }

          return (
            tokenBalance && (
              <BalanceContainer>
                <TokenBalance variant="title" size="small">
                  {tokenBalance.amount}
                </TokenBalance>
                <div />
                {tokenBalance.usdValue && (
                  <Typography
                    variant="body"
                    className={usdValueStyles()}
                    size="xsmall">
                    {`$${tokenBalance.usdValue}`}
                  </Typography>
                )}
              </BalanceContainer>
            )
          );
        };
        
```

then 
```
end={endRenderer()}
```

mikasackermn replied:
okay, I added endRenderer to token list component

yeager-eren reviewed (approved):
Thanks. lgtm.

Please also ask @Ikari-Shinji-re to confirm `swap` changes.
